### PR TITLE
#169257408 Integrate the requests table to the backend

### PR DESCRIPTION
**Pull request by b0nbon1**
#### What does this PR do?
Integrates table user's request with the backend. Fetches all user's requests, filter the requests by status and also fetches a specific request for the user.
#### Description of Task to be completed?
- [x] make the endpoint protected
- [x] add table requests
- [x] add endpoint get all requests
- [x] add endpoint get specific requests
- [x] add middleware filter request by status
#### How should this be manually tested?
- git clone the repo
- drop the database and then recreate it manually(some changes have been made on the migrations)
- run `npm run migrate` to create the tables
- sign up or sign in a user  and grab the token provided
- create a request using endpoint `POST  http://localhost:3000/api/v1/trips/oneway` or `POST http://localhost:3000/api/v1/trips/return` and provide the token of the signed-in user on `bearer auth`
- then go to endpoint get request and use `GET localhost:3000/api/v1/requests` using the same user that created a request.
- also, go-to endpoint Get request and use `GET localhost:3000/api/v1/requests/:id` using the same user that created a request.
- go to endpoint get request and use `GET localhost:3000/api/v1/requests?status=open`, here the we filter the request using status which are `'open', 'approved', 'declined'`. using the same user that created a request.
#### Any background context you want to provide?
- approving or decline request hasn't been implemented yet, so filtering is for status `open`currently
#### What are the relevant pivotal tracker stories?
[#169257408](https://www.pivotaltracker.com/n/projects/2407417/stories/169257408)
#### Screenshots:
*Fetch all Requests:*

image

*Fetch all Requests with status open:*

image

*Fetch all Requests with status approved:*

image

*Fetch single Request:*

image

### CODE REVIEW

**Review by gildniy — DISMISSED** [review]
Tested and working.

File: src/controllers/request.controller.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+  /**
+   *
+   * @param {*} req
+   * @param {*} res
+   * @returns {object} succesful fetched the requests
+   */
+  async getOne(req, res) {
+    const currentUser = res.locals.user;
+    const requests = await db.request.findAll({
```
Comment (timolinn) [inline src/controllers/request.controller.js]: I think we can use `findOne` instead?

File: src/controllers/trips.controller.js [hunk 2]
```diff
@@ -35,15 +36,16 @@ class Trips {
...
       goingTo,
       travelDate,
       returnDate,
       reason,
       hotelId,
+      requestId,
       userId: currentUser.userId,
-      status: 'pending',
       type: 'return'
     });
```
Comment (timolinn) [inline src/controllers/trips.controller.js]: Why not create a trip service and wrap this in one of it's methods?

File: src/tests/scripts/factoties.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import db from '../../models';
```
Comment (timolinn) [inline src/tests/scripts/factoties.js]: Did you mean to name this file `factories`?

File: src/controllers/trips.controller.js (hunk 2 above)
Comment (b0nbon1) [reply src/controllers/trips.controller.js]: I think that is a better way to organizer the controller. Let me work on that thank you.

File: src/tests/scripts/factoties.js (hunk 3 above)
Comment (b0nbon1) [reply src/tests/scripts/factoties.js]: Yeah, yeah let me fix it

File: src/controllers/request.controller.js (hunk 1 above)
Comment (b0nbon1) [reply src/controllers/request.controller.js]: Oh, yeah on that.

File: src/controllers/request.controller.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import Responses from '../utils/response';
+import db from '../models';
+
+/**
+ * Class requests
+ */
+class Requests {
+  /**
```
Comment (nshaphan) [inline src/controllers/request.controller.js]: Well done @b0nbon1, but I think you would also move these to request service

File: src/controllers/request.controller.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  async getAll(req, res) {
+    const currentUser = res.locals.user;
+    const requests = await db.request.findAll({
+      where: {
+        userId: currentUser.userId
+      },
+      include: [{
+        model: db.trip
+      }]
+    });
```
Comment (nshaphan) [inline src/controllers/request.controller.js]: Well done @b0nbon1, but I think you would also move these to request service

File: src/controllers/request.controller.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    const currentUser = res.locals.user;
+    const requests = await db.request.findOne({
+      where: {
+        userId: currentUser.userId,
+        id: req.params.id
+      },
+      include: [{
+        model: db.trip
+      }]
+    });
```
Comment (nshaphan) [inline src/controllers/request.controller.js]: This also can be moved to request service

**Review by gildniy — APPROVED** [review]
Looks nice, I tested and it works.

**Review by nshaphan — APPROVED** [review]
Well done @b0nbon1

File: src/scripts/rollback.js [hunk 7]
```diff
@@ -10,7 +10,7 @@ const spawnOptions = { cwd: path.join(__dirname, '../..'), stdio: 'inherit' };
   const { url } = config.database;
 
   try {
-    await spawn('./node_modules/.bin/sequelize', ['db:migrate:undo', `--url=${url}`], spawnOptions);
+    await spawn('./node_modules/.bin/sequelize', ['db:migrate:undo:all', `--url=${url}`], spawnOptions);
```
Comment (RaymondMwaura) [inline src/scripts/rollback.js:13]: Great job on this, testing on both on Terminal and Postman will be much easier now.

**Review by RaymondMwaura — APPROVED** [review]
I think the PR description was sufficiently detailed. I have tested the feature with Postman as well as the unit and integration tests and all the functionality seems to be in order. Great job.

**Pull request merged**